### PR TITLE
Enterprise Search Engines PUT API

### PR DESCRIPTION
**Pull request by carlosdelest**
## Closes https://github.com/elastic/enterprise-search-team/issues/3928

Add PUT API for upserting engines:

```
PUT /_engine/test-engine
{
  "indices": ["test1", "test2"]
}
```

The API creates the Engine doc into the .engine system index, and adds an engine-<engine-name> alias for all indices.

As the alias is created, user needs to have manage permissions on both the alias and the indices.

Updating the engine indices updates the aliases as needed.

### CODE REVIEW

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/EnterpriseSearch.java [hunk 1]
```diff
@@ -85,7 +111,12 @@ public Collection<Object> createComponents(
         if (enabled == false) {
             return Collections.emptyList();
         }
-        final EngineIndexService engineService = new EngineIndexService(client, clusterService, namedWriteableRegistry, null);
+        final EngineIndexService engineService = new EngineIndexService(
+            client,
+            clusterService,
+            namedWriteableRegistry,
+            BigArrays.NON_RECYCLING_INSTANCE
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/EnterpriseSearch.java:118]: I couldn't find any other way of setting a BigArrays instance, is this OK?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java [hunk 2]
```diff
@@ -195,6 +195,7 @@ public void putEngine(Engine engine, ActionListener<IndexResponse> listener) {
                     )
                     .endObject();
             }
+            // TODO Validate indices exist?
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java:198]: I was thinking on validating that the indices exist here. WDYT?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        public Request(String engineId, BytesReference content, XContentType contentType) {
+            this.engine = Engine.fromXContentBytes(engineId, content, contentType);
+            xContentType = contentType;
+        }
+
+        @Override
+        public ActionRequestValidationException validate() {
+            ActionRequestValidationException validationException = null;
+
+            if (engine.indices().length == 0) {
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java:60]: I guess at this point we'll just do syntactic validation (indices have been provided in the request) and afterwards we can check whether indices exist.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/RestPutEngineAction.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import org.elasticsearch.rest.action.RestToXContentListener;
+
+import java.io.IOException;
+import java.util.List;
+
+import static org.elasticsearch.rest.RestRequest.Method.PUT;
+
+public class RestPutEngineAction extends BaseRestHandler {
+
+    public static final String ENDPOINT = "_engine";
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/RestPutEngineAction.java:22]: There'll be probably a shared ENDPOINT constant for all endpoints

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        super(NAME, PutEngineAction.Response::new);
+    }
+
+    public static class Request extends ActionRequest {
+
+        private final Engine engine;
+        private XContentType xContentType = XContentType.JSON;
+
+        public Request(StreamInput in) throws IOException {
+            super(in);
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java:43]: I'm very interested in knowing whether this way of handling the request and extracting the Engine is correct.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationUtils.java [hunk 6]
```diff
@@ -28,6 +28,7 @@
 import static org.elasticsearch.xpack.core.ClientHelper.ASYNC_SEARCH_ORIGIN;
 import static org.elasticsearch.xpack.core.ClientHelper.DEPRECATION_ORIGIN;
 import static org.elasticsearch.xpack.core.ClientHelper.ENRICH_ORIGIN;
+import static org.elasticsearch.xpack.core.ClientHelper.ENT_SEARCH_ENGINE_ORIGIN;
```
Comment (ioanatia) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationUtils.java:31]: could we make this just `ENT_SEARCH`? we might need it for more than just engines, e.g. analytics, connector related features

Comment (carlosdelest) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationUtils.java:31]: That's a good point. Should we create multiple origins for that, or use a single one?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ClientHelper.java [hunk 7]
```diff
@@ -189,6 +189,7 @@ private static String maybeRewriteSingleAuthenticationHeaderForVersion(
     public static final String SEARCHABLE_SNAPSHOTS_ORIGIN = "searchable_snapshots";
     public static final String LOGSTASH_MANAGEMENT_ORIGIN = "logstash_management";
     public static final String FLEET_ORIGIN = "fleet";
+    public static final String ENT_SEARCH_ENGINE_ORIGIN = "engine";
```
Comment (kderusso) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ClientHelper.java:192]: It's going to be a race to see which one of us commits this first 😀

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/EnterpriseSearch.java [hunk 8]
```diff
@@ -63,7 +71,25 @@ protected XPackLicenseState getLicenseState() {
...
+        IndexScopedSettings indexScopedSettings,
+        SettingsFilter settingsFilter,
+        IndexNameExpressionResolver indexNameExpressionResolver,
+        Supplier<DiscoveryNodes> nodesInCluster
+    ) {
+
+        if (enabled == false) {
+            return Collections.emptyList();
+        }
+        // Register new actions here
```
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/EnterpriseSearch.java:91]: I think we can remove this comment now that we have actions to register.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/EnterpriseSearch.java (hunk 1 above)
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/EnterpriseSearch.java:118]: This looks consistent with how its done elsewhere in the code base.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java (hunk 2 above)
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java:198]: Would it be better to put that validation in the `TransportPutEngineAction`?

It feels to me that the API layer is the place we want to do this validation and contract enforcement, not in the service layer. This way we can also ensure that the API caller has permissions on the individual indices if so desired. WDYT?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/RestPutEngineAction.java (hunk 4 above)
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/RestPutEngineAction.java:22]: ➕

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/TransportPutEngineAction.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+import org.elasticsearch.action.index.IndexResponse;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.entsearch.engine.Engine;
+import org.elasticsearch.xpack.entsearch.engine.EngineIndexService;
+
+public class TransportPutEngineAction extends HandledTransportAction<PutEngineAction.Request, PutEngineAction.Response> {
```
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/TransportPutEngineAction.java:20]: I think we want to do API request contract validation here? Examples might include: 

- Indices exist
- Requesting user has permissions on all indices 
- Indices are open? (This is outside the scope of the Rails API)
- etc...

**Review by kderusso — COMMENTED** [review]
Looks great so far @carlosdelest ! Looking forward to seeing the tests and additional validation. When you get a free moment could you also please update the PR description? Thanks!

**Review by jimczi — COMMENTED** [review]
Thanks @carlosdelest , I left some comments and questions.
For testing, yml tests should be fine. Let's talk offline how to setup them in the module.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/EnterpriseSearch.java (hunk 1 above)
Comment (DaveCTurner) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/EnterpriseSearch.java:118]: It's not ok for the finished product, we should use the real `BigArrays` which recycles pages. But it is ok for now.

**Review by DaveCTurner — COMMENTED** [review]
The name "engine" is kinda generic, and there is already (at least) one other major concept in Elasticsearch called an Engine. Endpoint URIs are effectively forever: they're very very hard to change without breaking backwards compatibility. I think it's worth having a fairly broad discussion about the URI structure of this new functionality before we commit to something that we might regret in future. Sorry if there has already been such a discussion and I've just missed it.

Also (less critical to resolve up-front) are we sure that this should go in `x-pack.plugin.core` as opposed to a separate module?

I left a another couple of tiny comments.

**Comment by DaveCTurner** [comment]
> Also (less critical to resolve up-front) are we sure that this should go in x-pack.plugin.core as opposed to a separate module?

Apologies, I misread the diff, this is already not in core (except for one thing)

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java (hunk 5 above)
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java:42]: Makes sense, changed it!

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/TransportPutEngineAction.java [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+        engineIndexService.putEngine(engine, new ActionListener<>() {
+            @Override
+            public void onResponse(IndexResponse indexResponse) {
+                listener.onResponse(new PutEngineAction.Response(indexResponse.getResult()));
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                listener.onFailure(e);
+            }
```
[the remark this answers is not in the document]
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/TransportPutEngineAction.java:42]: TIL, thanks @DaveCTurner !

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/EnterpriseSearch.java (hunk 1 above)
Comment (jimczi) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/EnterpriseSearch.java:118]: > should we keep it as a Component, to avoid having to explicitly create it in the TransportActions?

It's just a client wrapper so I think it's ok to create it on each TransportAction for now.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java [hunk 11]
```diff
@@ -177,13 +186,68 @@ public void getEngine(String resourceName, ActionListener<Engine> listener) {
         }));
     }
 
+    private static String getEngineAliasName(Engine engine) {
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java:189]: This could be part of the Engine class, but I feel it's strongly related to how this service works and not the data that is held by the Engine.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java [hunk 12]
```diff
@@ -177,13 +186,68 @@ public void getEngine(String resourceName, ActionListener<Engine> listener) {
...
+    }
+
     /**
      * Creates or updates the {@link Engine} in the underlying index.
      *
      * @param engine The engine object.
      * @param listener The action listener to invoke on response/failure.
      */
     public void putEngine(Engine engine, ActionListener<IndexResponse> listener) {
+        validateIndices(engine);
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java:200]: I opted for explicitly validating the indices. Otherwise, creating the alias returns a 404 - which is weird, given that it's an error on the user params. This way we can control what we return to the user.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java [hunk 13]
```diff
@@ -177,13 +186,68 @@ public void getEngine(String resourceName, ActionListener<Engine> listener) {
...
+
+    }
+
+    private void createOrUpdateAlias(Engine engine, ActionListener<AcknowledgedResponse> listener) {
+
+        final Metadata metadata = clusterService.state().metadata();
+        final String engineAliasName = getEngineAliasName(engine);
+
+        IndicesAliasesRequestBuilder requestBuilder = null;
+        if (metadata.hasAlias(engineAliasName)) {
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java:212]: A few cases we need to check:
- The alias has not yet been created
- Alias is created, and we need to update the indices
- Alias is created, and there's nothing to do. We don't do an operation or validation would fail as there is no action.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java [hunk 14]
```diff
@@ -177,13 +186,68 @@ public void getEngine(String resourceName, ActionListener<Engine> listener) {
...
+            Set<String> currentAliases = metadata.aliasedIndices(engineAliasName)
+                .stream()
+                .map(index -> index.getName())
+                .collect(Collectors.toSet());
+            Set<String> targetAliases = Set.of(engine.indices());
+
+            if (currentAliases.equals(targetAliases)) {
```
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java:219]: We came up with a subtle edge case when developing this in the Rails app: 

- User creates an Engine with `index1` and `index2`. 
- User deletes `index2`. It is automatically removed from the alias. 
- User re-creates `index2`. The alias is not automatically updated. 
- User adds `index2` to the engine. Since `index2` was already associated with the engine no action is taken and the index is not added back to the alias. 

The workaround here I think is to always add all the indices to the alias on every PUT request. It's a cheap and atomic operation.

**Comment by carlosdelest** [comment]
Nice, thanks for the tip! I changed the code so aliases are always re-added.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java [hunk 15]
```diff
@@ -177,13 +186,64 @@ public void getEngine(String resourceName, ActionListener<Engine> listener) {
...
+    }
+
+    private IndicesAliasesRequestBuilder updateAliasIndices(Set<String> currentAliases, Set<String> targetAliases, String engineAliasName) {
+
+        Set<String> deleteIndices = new HashSet<>(currentAliases);
+        deleteIndices.removeAll(targetAliases);
+
+        IndicesAliasesRequestBuilder aliasesRequestBuilder = clientWithOrigin.admin().indices().prepareAliases();
+
+        // Always re-add aliases, as an index could have been removed manually and it must be restored
```
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java:235]: I apologize, I made that comment and then deleted it 🤦 I don't know if this is needed since you're pulling from the cluster metadata rather than the engine. Still better safe than sorry I suppose.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java [hunk 16]
```diff
@@ -215,6 +284,7 @@ public void putEngine(Engine engine, ActionListener<IndexResponse> listener) {
      */
     public void deleteEngine(String engineName, ActionListener<DeleteResponse> listener) {
         try {
+            // TODO Delete alias when Engine is deleted
```
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java:287]: I've added this to the acceptance criteria of the followup issue, this comment is no longer needed. Thank you!

**Review by kderusso — APPROVED** [review]
Overall changes LGTM. A few comments but nothing that immediately prevents merging into the branch.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+import org.elasticsearch.xpack.entsearch.engine.Engine;
+
+import java.io.IOException;
+
+import static org.elasticsearch.action.ValidateActions.addValidationError;
+
+public class PutEngineAction extends ActionType<PutEngineAction.Response> {
+
+    public static final PutEngineAction INSTANCE = new PutEngineAction();
+    public static final String NAME = "cluster:admin/engine/put";
```
Comment (jimczi) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java:30]: Since we require the manage index privilege for the creation of the alias we should use `"indices:admin/engine/put"`. That way we don't require another privilege for engine management. Managements of engines should be done under the manage index privilege. 
The advantage is  that we get the resolution of index pattern (`index-*`) to respect the underlying security.
Indices resolution happens at the transport level on the `IndicesRequest`. So instead of `ActionRequest`, the `Request` below should extends `IndicesRequest` and get/set the list indices from the `IndicesRequest::indices` function.

Comment (jimczi) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java:30]: Also note that the tests are failing due to:
```
 Actions are neither operator-only nor non-operator: [[cluster:admin/engine/put]]. An action should be declared to be either operator-only in [org.elasticsearch.xpack.security.operator.OperatorOnlyRegistry] or non-operator in [org.elasticsearch.xpack.security.operator.Constants
```
The new action needs to be added in `org.elasticsearch.xpack.security.operator.Constant`

Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java]: Thanks Jim! I've implemented IndicesRequest in the Request. I'd like some advice on the `IndicesOptions` - I guess we don't want any expansions, so I've set to `false` all options. Is there some better way to do this, or should we allow any of the `IndicesOptions`?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java [hunk 18]
```diff
@@ -205,6 +265,15 @@ public void putEngine(Engine engine, ActionListener<IndexResponse> listener) {
         }
     }
 
+    private void validateIndices(Engine engine) {
+        for (String index : engine.indices()) {
+            final IndexMetadata indexMetadata = clusterService.state().metadata().index(index);
+            if (indexMetadata == null) {
```
[the remark this answers is not in the document]
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java:271]: yep, I thought about that too. I think this way it is consistent with other ES behaviour (for example, searching on multiple indices) and we simplify this code as well.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java (hunk 15 above)
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java:235]: I think it doesn't hurt - updating an engine shouldn't be that frequent, we can risk doing an additional update alias even if not needed.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java (hunk 11 above)
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java:189]: It's two methods as of now - let's discuss this when we add the delete part 👍

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+public class PutEngineAction extends ActionType<PutEngineAction.Response> {
+
+    public static final PutEngineAction INSTANCE = new PutEngineAction();
+    public static final String NAME = "indices:admin/engine/put";
+
+    public PutEngineAction() {
+        super(NAME, PutEngineAction.Response::new);
+    }
+
+    public static class Request extends ActionRequest implements IndicesRequest {
```
[the remark this answers is not in the document]
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java:38]: OK, I've done that, please check the implementation 🙏

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+        public IndicesOptions indicesOptions() {
+            return IndicesOptions.fromOptions(false, false, false, false, false, false, false, false);
+        }
+
+        public Engine getEngine() {
+            return engine;
+        }
+    }
+
+    public static class Response extends ActionResponse implements ToXContentObject {
```
[the remark this answers is not in the document]
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java:83]: OK! I've also checked the status code for the `PutEngineRestAction`, so it actually returns this status instead of 200 all the time. Please check!

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (jimczi) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java:38]: I don't see the change. What I suggest is to do:
`public static class Request extends ActionRequest implements IndicesRequest.Replaceable {`
and then add an implementation for `indices(String.. indices)` and `String indices()`.

**Review by jimczi — COMMENTED** [review]
Can you also add tests for the request/response serialisation similar to this [test](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/autoscaling/src/test/java/org/elasticsearch/xpack/autoscaling/action/GetAutoscalingPolicyActionResponseWireSerializingTests.java) ?

**Comment by carlosdelest** [comment]
Closing in favour of https://github.com/elastic/elasticsearch/pull/94072 - this PR got unlinked from the original branch somehow, we can't see updates to the branch reflected here 🤷

**Comment by carlosdelest** [comment]
I'm stupid and was doing some work on other branch already - restoring this original PR. Sorry about that folks.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java [hunk 21]
```diff
@@ -0,0 +1,129 @@
...
+        }
+
+        @Override
+        public IndicesRequest indices(String... indices) {
+            return new Request(new Engine(engine.name(), indices, engine.analyticsCollectionName()));
+        }
+
+        @Override
+        public IndicesOptions indicesOptions() {
+            return IndicesOptions.fromOptions(false, false, false, false, false, false, false, false);
```
Comment (jimczi) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java:85]: I don't think we enable the right options here. We should copy what is done for aliases [here](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/admin/indices/alias/IndicesAliasesRequest.java#L60).

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java [hunk 22]
```diff
@@ -0,0 +1,129 @@
...
+            builder.endObject();
+            return builder;
+        }
+
+        @Override
+        public RestStatus status() {
+            return switch (result) {
+                case CREATED -> RestStatus.CREATED;
+                case NOT_FOUND -> RestStatus.NOT_FOUND;
+                default -> RestStatus.OK;
```
Comment (jimczi) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java:124]: Why do we need to restrict the status code? It should be possible to use this action to override an existing Engine, no?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (jimczi) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java:271]: Are we accepting index patterns? If so, this won't work with this validation. I also wonder why we need this validation since the alias action checks the existence of the indices already.

File: x-pack/plugin/ent-search/build.gradle [hunk 23]
```diff
@@ -19,3 +19,8 @@ dependencies {
 }
 
 addQaCheckDependencies(project)
+
+testClusters.configureEach {
+  setting 'xpack.ent-search.enabled', 'true'
+}
```
Comment (jimczi) [inline x-pack/plugin/ent-search/build.gradle:25]: Why is this needed?

**Review by jimczi — COMMENTED** [review]
It's getting close, thanks @carlosdelest 
I left some additional comments. Let's make sure we have the tests implemented before merging.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java (hunk 17 above)
Comment (jimczi) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java]: I added a comment here: https://github.com/elastic/elasticsearch/pull/93975#discussion_r1115580241

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java (hunk 22 above)
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java:124]: Yes, in that case the status will be OK - there's no specific status for an update AFAIU

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java (hunk 21 above)
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java:85]: I'm fine with that. Do we want to allow wildcard expansions? Would it be possible then to issue the following request?

PUT /_engine/test-engine-another-one
{
  "indices": ["test*"]
}

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java (hunk 22 above)
Comment (jimczi) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java:124]: Ok cool, thanks

File: x-pack/plugin/ent-search/build.gradle (hunk 23 above)
Comment (carlosdelest) [reply x-pack/plugin/ent-search/build.gradle:25]: I guess we will need the feature toggle enabled to perform the API testing - I included it here for the subsequent testing work, but I can remove it and add it later.

Comment (jimczi) [reply x-pack/plugin/ent-search/build.gradle:25]: The config for the test clusters is [here](https://github.com/elastic/elasticsearch/blob/ent-search-module/x-pack/plugin/ent-search/qa/rest/build.gradle). 
We don't have rest tests outside of `qa`.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java:271]: > Are we accepting index patterns?

I was thinking that no, until we added IndicesRequest.Replaceable to the Request. Wouldn't that take care of index patterns and replace them on the request?

>  I also wonder why we need this validation since the alias action checks the existence of the indices already

That returned a not found response from the alias, and was a bit confusing since we were returning a 404 directly. I could capture that result from the alias operation and issue a 400 response to make clear we were trying to use a non-existing index.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java (hunk 21 above)
Comment (jimczi) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java:85]: Was it allowed in the rails app? If not we can just disable index pattern resolution in the IndexOptions.

File: x-pack/plugin/ent-search/build.gradle (hunk 23 above)
Comment (carlosdelest) [reply x-pack/plugin/ent-search/build.gradle:25]: Got it, removing that!

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java (hunk 21 above)
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java:85]: This was not allowed in the Rails app. I'll disable index pattern resolution.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java (hunk 22 above)
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java:138]: Why do we need to return a 404 for the PUT command?

File: rest-api-spec/src/main/resources/rest-api-spec/api/engines.put.json [hunk 24]
```diff
@@ -0,0 +1,39 @@
+{
```
Comment (carlosdelest) [inline rest-api-spec/src/main/resources/rest-api-spec/api/engines.put.json:1]: Added REST tests, can you please check @jimczi ?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java (hunk 22 above)
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java:152]: It depends on what the response issued from us would be. I just didn't want a 404 result to pass through undetected - for example, it could happen in the alias response in case it is not handled properly for changing that into a 400.

File: x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineActionRequestSerializingTests.java [hunk 25]
```diff
@@ -0,0 +1,41 @@
...
+package org.elasticsearch.xpack.entsearch.engine.action;
+
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.test.AbstractWireSerializingTestCase;
+import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.xpack.entsearch.engine.Engine;
+
+import java.io.IOException;
+
+public class PutEngineActionRequestSerializingTests extends AbstractWireSerializingTestCase<PutEngineAction.Request> {
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineActionRequestSerializingTests.java:17]: Added serializing test cases, can you please check @jimczi ?

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_put.yml [hunk 26]
```diff
@@ -0,0 +1,72 @@
+setup:
```
Comment (jimczi) [inline x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_put.yml:1]: nit: Let's call the file 20_engine.yml or  20_engine_basic to leave room for testing the get and remove method?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/EnterpriseSearch.java (hunk 1 above)
Comment (jimczi) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/EnterpriseSearch.java:118]: Can you add a TODO here so that we don't forget?

File: rest-api-spec/src/main/resources/rest-api-spec/api/engines.put.json [hunk 27]
```diff
@@ -0,0 +1,39 @@
...
+  "engines.put": {
```
Comment (jimczi) [inline rest-api-spec/src/main/resources/rest-api-spec/api/engines.put.json:2]: let's call it engine singular to match the URI?

**Review by jimczi — APPROVED** [review]
The PR looks great @carlosdelest . I left two nit comments but feel free to merge if the tests are green!

File: rest-api-spec/src/main/resources/rest-api-spec/api/engines.put.json (hunk 27 above)
Comment (carlosdelest) [reply rest-api-spec/src/main/resources/rest-api-spec/api/engines.put.json]: Done!

**Comment by kderusso** [comment]
elasticmachine run elasticsearch-ci/bwc
